Update deps and msrv

vincentdephily opened the pull request:
What started as a wish to fix #20 ended up as a larger deps and MSRV update (#22), plus fixing some bitrot. Nothing too scary, but the MSRV bump does imply a semver bump.

Rust 1.16 was 11 months old when it was chosen as the MSRV. The new MSRV (rust 1.32) is 16 months old and brings rust edition 2018, so it should have wide adoption at this stage.

I kept an eye on build/test/bench using both the MSRV and nightly compiler, as well as minimal dep versions. In the CI, I use minimal-versions with nightly only because I don't know whether `cargo +nightly update -Z minimal-version`  would succeed when `$TRAVIS_RUST_VERSION = 1.32.0`. If that's supported, it would be a better CI setup.

Thanks for the review.

cardoe commented:
So fwiw, I don't pick the MSRV. Because this crate ties in with crates.io/crates/log we used that. Until very recently, that was 1.16.0 which you can see in https://github.com/rust-lang/log/commit/433731637477b27f573a3e40cd2297c5a776ff1b

However, the MSRV needs to be 1.31.0 because that's what's currently set per https://github.com/rust-lang/log/commit/ae59aa14de647920eec512d0b02661506c1a3609 so we cannot move all the way up to 1.32.0

cardoe commented on examples/docopt.rs:
```diff
@@ -6,29 +6,25 @@
 use docopt::Docopt;
+use log::*;
 use std::env;
+use serde::Deserialize;
```
This doesn't work because of how serde was imported in Cargo.toml.

cardoe commented on Cargo.toml:
```diff
@@ -20,15 +20,16 @@ is-it-maintained-open-issues = { repository = "cardoe/stderrlog-rs" }
+clap = "~2.33.0"
+docopt = "1.1.0"
+serde = "1.0"
+serde_derive = "1.0"
```
You got rid of the extern crates but then didn't import serde appropriately.
`serde = { version = "1.0", features = ["derive"] }` and getting rid of `serde_derive`

vincentdephily replied:
Looks like you fixed that in your version, thanks.

cardoe commented on CHANGELOG.md:
```diff
@@ -1,5 +1,9 @@
 
+## 0.5.0 (unreleased)
+
+- bump minimum Rust version to 1.32.0
```
Gotta stick with 1.31.0 to stay compatible with https://crates.io/crate/log

vincentdephily replied:
Yup, happy to stick with 1.31 to match `log`'s MSRV. We'll need to fix our docs to mention the requirement for `log >= 0.4.10` instead of `0.4.1`.

vincentdephily replied:
Latest commit should have updated all the docs.

cardoe commented on .travis.yml:
```diff
@@ -5,16 +5,14 @@ rust:
-      cargo update -p cfg-if --precise 0.1.9;
+  - if [ ${TRAVIS_RUST_VERSION} = "nightly" ]; then
+      # Ideally we would like to use minimal-versions with our MSRV, but this is a nightly-only option.
+      cargo generate-lockfile -Z minimal-versions
```
I wrote something similar to this cause of the `-Z mininal-versions` PR but this wouldn't work correctly because the steps below didn't do `--locked`, which then would require doing it to the non-nightly.

vincentdephily replied:
Your version of this change look good, thanks.

Does travis give access to all versions though ? It would be better to test minimal-versions with the MSRV. Something like

```shell
   if [ ${TRAVIS_RUST_VERSION} = "1.31.0" ]; then
      cargo +nightly generate-lockfile -Z minimal-versions;
    else
      cargo generate-lockfile;
    fi
```

But I'm a travis noob, so I'll let you experiment.

vincentdephily replied:
I went ahead and tried the change. Just don't cherry-pick it if travis doesn't like it.

cardoe commented on benches/common/mod.rs:
```diff
@@ -1,9 +1,10 @@
+use super::{init_logger, util::with_redirected_stderr};
+use log::*;
 use std::panic;
 use test::Bencher;
```
Some of these changes now don't pass a `cargo fmt` cleanly.

vincentdephily replied:
I thought I noticed that the original code wasn't fmt-clean, so I didn't pay too much attention to it. I'll setup a git hook to check for fmt.

vincentdephily commented on benches/junk_modules.rs:
```diff
@@ -3,9 +3,9 @@
+use std::sync::Once;
 
-static INIT_LOGGER: sync::Once = sync::ONCE_INIT;
+static INIT_LOGGER: Once = Once::new();
```
Yes, that requires 1.32 and is only cosmetic, so we can leave it out.

vincentdephily commented:
Thanks for detailed review. I rebased this branch on master and added a few commits:

- The atty version relax somehow got lost during cherry-picking
- Wanted to up the minimum log version to the one that raised the MSRV, but they are [currently yanked](https://github.com/rust-lang/log/issues/385) so I'm not sure what to do. On the one hand, stderrlog seems to work fine with 0.4.1. On the other hand, bumping the stderrlog semver is a good opportunity to bump the log crate version. It seems prudent to wait for log-1.4.11 before releasing stderrlog-0.5.0.
- Made the mentioned travis change, will see if it works. Might be nice to put a travis link in the README ;)

vincentdephily commented:
Log 0.4.11 was released two weeks ago, so I presume we can go ahead with merging this and doing our own release. I've added a commit to that effect, thanks for taking a look.

vincentdephily commented:
@cardoe Ping.

Is there anything blocking this merge and a release ? #20 has been fixed a while ago, but until we push an stderrlog release to crates.io, we'll still get dependency resolution failures.
